pkg/backends/calico/routes.go

```diff
@@ -190,29 +190,156 @@ func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
 			return
 		}
 	}
-	logc := log.WithField("svc", fmt.Sprintf("%s/%s", svc.Namespace, svc.Name))
 
 	// see if any endpoints are on this node and advertise if so
 	// else remove the route if it also already exists
 	rg.Lock()
-	if rg.advertiseThisService(svc, ep) {
-		route := svc.Spec.ClusterIP + "/32"
-		if cur, exists := rg.svcRouteMap[key]; !exists {
-			// This is a new route - send it through.
-			rg.advertiseRoute(key, route)
-		} else if route != cur {
-			// The route has changed somehow. Send a delete for the old one
-			// and add the new one. We don't expect to hit this, since cluster IPs
-			// are immutable. But, handle it for good measure.
-			logc.Warn("ClusterIP for service changed, adjusting")
-			rg.withdrawRoute(key, cur)
+	defer rg.Unlock()
+
+	advertise := rg.advertiseThisService(svc, ep)
+	if advertise {
+		routes := rg.getAllRoutesForService(svc)
+		rg.advertiseRoutes(key, routes)
+	} else {
+		routes := rg.getAdvertisedRoutes(key)
+		rg.withdrawRoutes(key, routes)
+	}
+
+}
+
+// onBGPConfigurationUpdate is called from the calico client, whenever there
+// is a change to the svc_external_ips, and will update all of the routes
+// advertised for each service.
+func (rg *routeGenerator) onBGPConfigurationUpdate() {
+
+	// Get all the services that we know about
+	svcs := make([]*v1.Service, 0)
+	svcIfaces := rg.svcIndexer.List()
+	for _, svcIface := range svcIfaces {
+		svc, ok := svcIface.(*v1.Service)
+		if !ok {
+			log.Error("Type assertion failed for rg.svcIndexer result member. Will not process updates to routes advertised for service.")
+			continue
+		}
+
+		svcs = append(svcs, svc)
+	}
+
+	// Update the routes advertised for each service
+	for _, svc := range svcs {
```

Comment:
Any reason we need the second loop here rather than just calling this for each service in turn in the loop above?

Reply:
Just to make the intent clearer - but you're right, it's unnecessary. I'll remove it.